### PR TITLE
RF: Simplify EnsureDataset and add optional ID check

### LINKED ISSUE #272: Add a dataset ID check for EnsureDataset?
In the context of metalad, I came across other custom checks that may make sense to integrate into existing Constraints. One custom additional check metalad does is a check for a valid dataset ID. Could ``EnsureDataset()`` get a check like this? Or should the constraints not get overloaded with options?

### PR DESCRIPTION
**Pull request by adswa**
This sits on top of #271 to prevent a merge conflict and make use of the changes required for type checking already implemented in it.

This PR adds two changes in individual commits:

The first commit refactors `EnsureDataset` - there was some duplication that I thought could be reduced by factoring out a ``_require_dataset()`` helper.

The second commit adds an optional `idcheck` to `EnsureDataset` to fix #272

### CODE REVIEW

File: datalad_next/constraints/dataset.py [hunk 1]
```diff
@@ -50,7 +41,10 @@ class EnsureDataset(Constraint):
     the DataLad core package). With ``installed=False`` no exception is
     raised and a dataset instances matching PWD is returned.
     """
-    def __init__(self, installed: bool = None, purpose: str = None):
+    def __init__(self,
+                 installed: bool | None = None,
+                 purpose: str | None = None,
+                 idcheck: bool | None = None):
```
Comment (mih) [inline datalad_next/constraints/dataset.py:47]: What about `with_id: bool`? It would read "ensure dataset with id", not "ensure dataset idcheck". The code is als not checking the ID, only whether it is more than nothing.

Maybe "with" have too much of meaning in the Python context. `require_id` would also do it for me.

File: datalad_next/constraints/dataset.py [hunk 2]
```diff
@@ -60,54 +54,64 @@ def __init__(self, installed: bool = None, purpose: str = None):
...
+                                 f'datalad-id')
         return DatasetParameter(value, ds)
 
     def short_description(self) -> str:
         return "(path to) {}Dataset".format(
             'an existing ' if self._installed is True
             else 'a non-existing ' if self._installed is False else 'a ')
+
+
+def _require_dataset(self, value):
```
Comment (mih) [inline datalad_next/constraints/dataset.py:96]: Better indent to make its nature as a method more obvious

File: datalad_next/constraints/tests/test_special_purpose.py [hunk 3]
```diff
@@ -356,3 +356,14 @@ def test_EnsureDataset(tmp_path):
...
+    # smoke test for idcheck:
+    assert EnsureDataset(idcheck=True)(ds).ds == ds
+    assert EnsureDataset(idcheck=False)(ds).ds == ds
+    # unset the dataset ID to test whether an ID check would raise, but
+    # bring it back later in case future tests need it
+    id = ds.config.get('datalad.dataset.id')
+    ds.config.unset('datalad.dataset.id')
+    with pytest.raises(NoDatasetFound):
+        EnsureDataset(idcheck=True)(tmp_path)
+    ds.config.set('datalad.dataset.id', id)
```
Comment (mih) [inline datalad_next/constraints/tests/test_special_purpose.py:369]: The dataset is temporary, there is no need to undo the changes. If you want to be more specific, you could use `scope='branch'` for `ds.config.unset()`.

If you want to be extra-safe re config leakage, use the `datalad_cfg` fixture. But all tests in `-next` are automatically verified to not change the persistent configuration. So if this one passes, we are safe.

**Review by mih — CHANGES_REQUESTED** [review]
Thanks, this works for me! I only left minor, and a naming suggestion. I could be convinced otherwise, though.

File: datalad_next/constraints/dataset.py (hunk 1 above)
Comment (adswa) [reply datalad_next/constraints/dataset.py:47]: I personally like ``require_id`` best

File: datalad_next/constraints/dataset.py (hunk 2 above)
Comment (adswa) [reply datalad_next/constraints/dataset.py:96]: thx

File: datalad_next/constraints/tests/test_special_purpose.py (hunk 3 above)
Comment (adswa) [reply datalad_next/constraints/tests/test_special_purpose.py]: thx for the tip!

File: datalad_next/constraints/dataset.py (hunk 1 above)
Comment (adswa) [reply datalad_next/constraints/dataset.py]: Renamed it to ``require_id`` now.

File: datalad_next/constraints/dataset.py (hunk 2 above)
Comment (adswa) [reply datalad_next/constraints/dataset.py]: done

File: datalad_next/constraints/tests/test_special_purpose.py (hunk 3 above)
Comment (adswa) [reply datalad_next/constraints/tests/test_special_purpose.py]: I've changed the scope to branch.

**Review by mih — APPROVED** [review]
LGTM! Thx!
